elki-core-math/src/main/java/elki/math/statistics/tests/mcde/MWPTest.java

```diff
@@ -0,0 +1,191 @@
+   * double[] corrected_ranks = corrected_ranks(...);
+   * double l = corrected_rank[0]; double adjusted_rank = corrected_rank[1]; double correction = corrected_rank[2];
+   * <p>
+   * Correspond to one instance of the original data
+   */
+  public <A> MwpIndex[] corrected_ranks(final NumberArrayAdapter<?, A> adapter, final A data, int[] idx) {
+    final int len = adapter.size(data);
+    MwpIndex[] I = (MwpIndex[]) Array.newInstance(MwpIndex.class, len);
+
+    int j = 0;
+    long correction = 0;
+    while(j < len) {
+      int k = j;
+      int t = 1;
+      double adjust = 0.0;
+
+      while((k < len - 1) && (adapter.getDouble(data, idx[k]) == adapter.getDouble(data, idx[k + 1]))) {
+        adjust += k;
+        k++;
+        t++;
+      }
+
+      if(k > j) {
+        double adjusted = (adjust + k) / t;
+        correction += (t * t * t) - t;
+
+        for(int m = j; m <= k; m++) {
+          I[m] = new MwpIndex(idx[m], adjusted, correction);
+        }
+      }
+      else
+        I[j] = new MwpIndex(idx[j], j, correction);
+      j += t;
+    }
+
+    return I;
+  }
+
+  /**
+   * Efficient implementation of MWP statistical test using appropriate index structure as described in Algorithm 3
+   * of reference paper. Tailored to MCDE Framework.
+   *
+   * @param start           Starting index value for statistical test
+   * @param end             End index value for statistical test
+   * @param slice           Return value of randomSlice() created with the index that is not for the reference dimension
+   * @param corrected_ranks Index of the reference dimension, return value of corrected_ranks() computed for reference dimension
+   * @return p-value from two sided Mann-Whitney-U test
+   */
+  public double statistical_test(int start, int end, boolean[] slice, MwpIndex[] corrected_ranks) {
+    double R = 0.0;
+    long n1 = 0;
+    for(int j = start; j < end; j++) {
+
+      if(slice[corrected_ranks[j].index]) {
+        R += corrected_ranks[j].adjusted;
+        n1++;
+      }
+    }
+
+    // This is to cancel the offset in case the marginal restriction does not start from 0
+    // see "acc - (cutStart * count)" is reference implementation of MWP
+    R -= start * n1;
+
+    final int cutLength = end - start;
+    if((n1 == 0) || (n1 == cutLength)) {
+      return 1;
+    }
+
+    final double U = R - ((double) (n1 * (n1 - 1))) / 2;
+    final long n2 = cutLength - n1;
+
+    if(n1 + n2 > two_times_sqrt_max_long) {
+      throw new AbortException("Long type overflowed. Too many objects: Please subsample and try again with smaller data set.");
+    }
+
+    final long b_end = corrected_ranks[(end - 1)].correction;
+    final long b_start = start == 0 ? 0 : corrected_ranks[(start - 1)].correction;
+    final double correction = (double) (b_end - b_start) / (cutLength * (cutLength - 1));
+    final double std = Math.sqrt((((double) (n1 * n2)) / 12) * (cutLength + 1 - correction));
+
+    if(std == 0) {
```

Comment:
Can this actually happen? Maybe this case could be caught earlier (`n1==0 || n2 == 0`)?

Reply:
Obviously it is important to not divide by 0 in line 178. Neither n1 nor n2 should get 0 so the only way this can get 0 is if correction == cutLength + 1. Today I could not really create a situation in which that happens and the only scenario I could intuitively imagine is if there is no variation in the data which can be ruled out as I tested it and it lead not to std being 0. I will talk with Eduard about it as he came up with this way of calculating the standard deviation and if necessary come back to you about it.

Reply:
I tested it and it is possible for `std` to be 0. What I did is I created a small number of random data (int) in a small range (0-10) (to enforce situations on which all data points in a slice are of the same value etc.) and let the parameter alpha and beta vary (so that the two slices also become uneven). As such I would suggest to add back 

```suggestion
    if(std == 0) {
    return 0; 
    }
```

The code of the test is: 
```suggestion
package elki.math.statistics.tests.mcde;

import elki.math.statistics.dependence.MCDEDependenceMeasure;
import elki.utilities.random.RandomFactory;

import java.util.Random;

public class Test {
  public static void main(String[] args){
    MWPTest mwp = MWPTest.STATIC;
    final int len = 10;
    final int value_bound = 10;

    double[] data_1 = new double[len];
    double[] data_2 = new double[len];
    RandomFactory rnd = new RandomFactory(0);
    final Random random = rnd.getSingleThreadedRandom();

    for(int j = 0; j < 10000; j++) {
      for(int i = 0; i < len; i++){
        data_1[i] = random.nextInt(value_bound);
        data_2[i] = random.nextInt(value_bound);
      }

      double beta = 0.1 + 0.9 * random.nextDouble();
      double alpha = 0.1 + 0.9 * random.nextDouble();

      MCDEDependenceMeasure mcde = new MCDEDependenceMeasure(1000, alpha, beta, rnd, mwp);
      mcde.dependence(data_1, data_2);
    }
  }
}
```

If you set an assertion or a conditional breakpoint in `MWPTest.statisticalTest()` you should be able to recreate it. Note that I have just created a `Test.java` file in the mcde [folder](https://github.com/elki-project/elki/tree/feature/mcde/elki-core-math/src/main/java/elki/math/statistics/tests/mcde) (where also `MCDETest` and `MWPTest` lie).